### PR TITLE
add provenance

### LINKED ISSUE #19: Add pipeline provenance
Add provenance as per: https://github.com/BCCDC-PHL/AMR-TB-amplicon-artic-nf

### PR DESCRIPTION
**Pull request by taranewman**
Fixes #19

### CODE REVIEW

File: modules/illumina.nf [hunk 1]
```diff
@@ -11,9 +11,22 @@ process performHostFilter {
 
     output:
     tuple val(sampleName), path("${sampleName}_hostfiltered_R1.fastq.gz"), path("${sampleName}_hostfiltered_R2.fastq.gz"), emit: fastqPairs
+    tuple val(sampleName), path("${sampleName}_performHostFilter_provenance.yml"), emit: provenance
 
     script:
     """
+    printf -- "- process_name: performHostFilter\\n"                                           >> ${sampleName}_performHostFilter_provenance.yml
+    printf -- "  tools:\\n"                                                                    >> ${sampleName}_performHostFilter_provenance.yml
+    printf -- "    - tool_name: bwa mem\\n"                                                    >> ${sampleName}_performHostFilter_provenance.yml
```
Comment (dfornika) [inline modules/illumina.nf]: Could you adjust to:

```yml
- tool_name: bwa
  subcommand: mem
```

...and:

```yml
- tool_name: samtools
  subcommand: sort
```

See this example on the plasmid-screen pipeline:

https://github.com/BCCDC-PHL/plasmid-screen/blob/2898065e1702c00d5b6acb38565f79ac81cd0633/modules/align_reads_to_reference_plasmid.nf#L24-L39

File: modules/illumina.nf [hunk 2]
```diff
@@ -102,10 +121,16 @@ process filterResidualAdapters {
...
     output:
-    tuple val(sampleName), path("*1_posttrim_filter.fq.gz"), path("*2_posttrim_filter.fq.gz")
+    tuple val(sampleName), path("*1_posttrim_filter.fq.gz"), path("*2_posttrim_filter.fq.gz"), emit: filter_out
+    tuple val(sampleName), path("${sampleName}_filterResidualAdapters_provenance.yml"), emit: provenance
 
     script:
     """
+    printf -- "- process_name: filterResidualAdapters\\n"                                                 >> ${sampleName}_filterResidualAdapters_provenance.yml
+    printf -- "  tools:\\n"                                                                               >> ${sampleName}_filterResidualAdapters_provenance.yml
+    printf -- "    - tool_name: filter_residual_adapters.py\\n"                                           >> ${sampleName}_filterResidualAdapters_provenance.yml
```
Comment (dfornika) [inline modules/illumina.nf:132]: This is interesting. In general we haven't been collecting provenance for custom scripts. I'm not opposed to it but we might think a bit about how we want to format. I like your approach of using a sha256 of the script as a version.

File: modules/illumina.nf [hunk 3]
```diff
@@ -171,9 +196,25 @@ process trimPrimerSequences {
     output:
     tuple val(sampleName), path("${sampleName}.mapped.bam"), path("${sampleName}.mapped.bam.bai"), emit: mapped
     tuple val(sampleName), path("${sampleName}.mapped.primertrimmed.sorted.bam"), path("${sampleName}.mapped.primertrimmed.sorted.bam.bai" ), emit: ptrim
+    tuple val(sampleName), path("${sampleName}_trimPrimerSequences_provenance.yml"), emit: provenance
 
     script:
     """
+    printf -- "- process_name: trimPrimerSequences\\n"                                             >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "  tools:\\n"                                                                        >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "    - tool_name: samtools\\n"                                                       >> ${sampleName}_trimPrimerSequences_provenance.yml
```
Comment (dfornika) [inline modules/illumina.nf:206]: I think we may want to list out each samtools subcommand as a separate tool. Something like:

```yml
tools:
  - tool_name: samtools
    subcommand: view
    parameters:
      ...
  - tool_name: samtools
    subcommand: index
    parameters:
      ...
  - tool_name: samtools
    subcommand: sort
    parameters:
      ...
```

**Review by dfornika — CHANGES_REQUESTED** [review]
Looking good. I have a few re-formatting suggestions. We should check that this conforms with our [Pipeline Provenance Schema](https://github.com/BCCDC-PHL/pipeline-provenance-schema). We may need to update the schema to capture provenance for custom scripts.

File: modules/illumina.nf (hunk 2 above)
Comment (taranewman) [reply modules/illumina.nf:132]: stolen from: @cabrerad :) https://github.com/BCCDC-PHL/AMR-TB-amplicon-artic-nf/blob/main/modules/illumina.nf#L133

Comment (dfornika) [reply modules/illumina.nf:132]: Ok, in the interest of getting this merged I think we can go with this. But I wonder if we might want to distinguish custom scripts that are packaged in the pipeline from 'tools' like bwa, samtools, etc.

File: README.md [hunk 4]
```diff
@@ -83,3 +83,87 @@ A script to do some basic QC is provided in `bin/qc.py`. It measures the % of re
 
 #### Output
 A subdirectory for each process in the workflow is created in `--outdir`. A `nml_upload` subdirectory containing dehosted fastq files and consensus sequences is included. 
+
+#### Provenance
+In the output directory for each sample, a provenance file will be written with the following format:
+```
+- process_name: performHostFilter
+  tools:
+    - tool_name: bwa mem
```
Comment (dfornika) [inline README.md]: Sorry this is a nit-pick but this should be:

```
- tool_name: bwa
```

File: README.md [hunk 5]
```diff
@@ -83,3 +83,87 @@ A script to do some basic QC is provided in `bin/qc.py`. It measures the % of re
...
+  tools:
+    - tool_name: filter_residual_adapters.py
+          sha256: c3d062687abf2bbec48721a562ec609742101eec82887b1f31b9997361da901e
+- process_name: trimPrimerSequences
+  tools:
+    - tool_name: samtools
+      tool_version: 1.10
+      subcommand: view
+      parameters:
+        - parameter: -F4
```
Comment (dfornika) [inline README.md]: I think this should be:

```
- parameter: -F
  value: 4
```

see: https://www.htslib.org/doc/samtools-view.html

File: modules/illumina.nf [hunk 6]
```diff
@@ -11,9 +11,23 @@ process performHostFilter {
 
     output:
     tuple val(sampleName), path("${sampleName}_hostfiltered_R1.fastq.gz"), path("${sampleName}_hostfiltered_R2.fastq.gz"), emit: fastqPairs
+    tuple val(sampleName), path("${sampleName}_performHostFilter_provenance.yml"), emit: provenance
 
     script:
     """
+    printf -- "- process_name: performHostFilter\\n"                                           >> ${sampleName}_performHostFilter_provenance.yml
+    printf -- "  tools:\\n"                                                                    >> ${sampleName}_performHostFilter_provenance.yml
+    printf -- "    - tool_name: bwa mem\\n"                                                    >> ${sampleName}_performHostFilter_provenance.yml
```
Comment (dfornika) [inline modules/illumina.nf]: Just:

```
- tool_name: bwa
```

...here

File: modules/illumina.nf [hunk 7]
```diff
@@ -171,9 +197,36 @@ process trimPrimerSequences {
...
 
     script:
     """
+    printf -- "- process_name: trimPrimerSequences\\n"                                             >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "  tools:\\n"                                                                        >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "    - tool_name: samtools\\n"                                                       >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "      tool_version: \$(samtools 2>&1 | grep "Version: " | cut -d ' ' -f 2)\\n"      >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "      subcommand: view\\n"                                                          >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "      parameters:\\n"                                                               >> ${sampleName}_trimPrimerSequences_provenance.yml
+    printf -- "        - parameter: -F4\\n"                                                        >> ${sampleName}_trimPrimerSequences_provenance.yml
```
Comment (dfornika) [inline modules/illumina.nf]: This should be:

```
- parameter: -F
  value: 4
```

File: modules/illumina.nf [hunk 8]
```diff
@@ -102,10 +122,16 @@ process filterResidualAdapters {
...
-    tuple val(sampleName), path("*1_posttrim_filter.fq.gz"), path("*2_posttrim_filter.fq.gz")
+    tuple val(sampleName), path("*1_posttrim_filter.fq.gz"), path("*2_posttrim_filter.fq.gz"), emit: filter_out
+    tuple val(sampleName), path("${sampleName}_filterResidualAdapters_provenance.yml"), emit: provenance
 
     script:
     """
+    printf -- "- process_name: filterResidualAdapters\\n"                                                 >> ${sampleName}_filterResidualAdapters_provenance.yml
+    printf -- "  tools:\\n"                                                                               >> ${sampleName}_filterResidualAdapters_provenance.yml
+    printf -- "    - tool_name: filter_residual_adapters.py\\n"                                           >> ${sampleName}_filterResidualAdapters_provenance.yml
+    printf -- "          sha256: \$(shasum -a 256 ${projectDir}/bin/filter_residual_adapters.py | cut -d ' ' -f 1)\\n"      >> ${sampleName}_filterResidualAdapters_provenance.yml
```
Comment (dfornika) [inline modules/illumina.nf]: I know this is a bit of a pain the way we implement writing this yaml using consecutive `printf` statements, but I think we need to be careful about whitespace and indentation to ensure that the output is a valid yaml file.

Could you adjust the indentation on the `sha256` line so the start of `sha256` aligns with the start of `tool_name` in the line above?

**Review by dfornika — CHANGES_REQUESTED** [review]
Looking good. I just have a few more suggested changes.

Sorry to be a bit nit-picky but it helps a lot with the parsing of these files if we get them to be as consistent as possible. Before we merge we should do a quick check with an example output to just make sure that `pyyaml` will parse it without throwing an error.

File: README.md (hunk 4 above)
Comment (taranewman) [reply README.md]: fixed!

File: README.md (hunk 5 above)
Comment (taranewman) [reply README.md]: fixed!

File: modules/illumina.nf (hunk 6 above)
Comment (taranewman) [reply modules/illumina.nf]: fixed!

File: modules/illumina.nf (hunk 7 above)
Comment (taranewman) [reply modules/illumina.nf]: fixed!

File: modules/illumina.nf (hunk 8 above)
Comment (taranewman) [reply modules/illumina.nf]: fixed!

**Review by dfornika — APPROVED** [review]
Looks good! Ready to merge.

**Pull request merged**